### PR TITLE
fastem: hardware configuration

### PR DESCRIPTION
**Pull request by philipwinklerdelmic**
Megafield, overview, and livestream acquisitions need different settings of the scanner. The new function sets these values based on keyword arguments which can be passed through a dictionary.

### CODE REVIEW

File: src/odemis/acq/fastem.py [hunk 1]
```diff
@@ -167,6 +167,75 @@ def __init__(self, name, coordinates):
...
+
+    # => compute the scale needed in X, use the same one in Y, and then compute the Y resolution.
+    # => set resolution to full FoV, and then adjust
+    scale = scanner.shape[0] / resolution[0]
+    scanner.scale.value = (scale, scale)
+    if scanner.resolution.value != TILE_RES:
+        logging.warning("Unexpected resolution %s on e-beam scanner, expected %s",
+                        scanner.resolution.value, TILE_RES)
+
+    md = scanner.getMetadata()
```
Comment (Sabrina85) [inline src/odemis/acq/fastem.py:228]: How does this work now. You deleted below the method that reads the rotation for multibeam acquisition on the scanner from the MD. Here, you now always set the single beam rotation right? Do you not need to pick the right MD depending on the mode (live, megafield acq or overview acq)?

File: src/odemis/acq/fastem.py [hunk 2]
```diff
@@ -552,7 +611,7 @@ def _run_overview_acquisition(f, stream, stage, area, live_stream):
     """
     :returns: (DataArray)
     """
-    _configure_overview_hw(stream.emitter)
```
Comment (Sabrina85) [inline src/odemis/acq/fastem.py:555]: You also need to drop the method itself in right?

File: src/odemis/acq/fastem.py [hunk 3]
```diff
@@ -167,6 +167,75 @@ def __init__(self, name, coordinates):
...
+                        scanner.resolution.value, TILE_RES)
+
+    md = scanner.getMetadata()
+    if model.MD_SINGLE_BEAM_ROTATION in md:
+        scanner.rotation.value = md[model.MD_SINGLE_BEAM_ROTATION]
+    else:
+        scanner.rotation.value = SINGLE_BEAM_ROTATION_DEFAULT
+        logging.warning("Scanner doesn't have SINGLE_BEAM_ROTATION metadata, using %s rad.",
+                        scanner.rotation.value)
+    md[model.MD_ROTATION_COR] = -scanner.rotation.value
```
Comment (Sabrina85) [inline src/odemis/acq/fastem.py:235]: This was only needed for overview image acquisition. If I remember correctly in order to display the overview image at the same angle as the megafield image (where we do not use MD_ROTATION_COR) in Odemis. So you would need to distinguish here between live/overview and megafield as well.
@pieleric Can you comment or confirm, so we can add a comment here for clarity?
Is this the MD that goes on the image?

File: src/odemis/acq/fastem.py [hunk 4]
```diff
@@ -167,6 +167,75 @@ def __init__(self, name, coordinates):
...
+}
+
+
+SCANNER_CONFIG_LIVESTREAM = {
+    "multibeam_mode": False,
+    "external": False,
+    "blanker": None,  # automatic
+    "immersion": True,
+    "horizontal_fov": 2.2e-5,
+    "resolution": (1024, 884),
```
Comment (Sabrina85) [inline src/odemis/acq/fastem.py:196]: Add a comment saying that resolution for megafield imaging is controlled by the acquisition server module (ASM).

**Review by Sabrina85 — CHANGES_REQUESTED** [review]

File: src/odemis/acq/fastem.py (hunk 1 above)
Comment (philipwinklerdelmic) [reply src/odemis/acq/fastem.py:228]: I don't know. Is this something that is different for each mode? If so, how?

File: src/odemis/acq/fastem.py [hunk 5]
```diff
@@ -167,6 +167,75 @@ def __init__(self, name, coordinates):
         # if None -> acquire, if not None, don't acquire again
 
 
+SCANNER_CONFIG_OVERVIEW = {
+    "multibeam_mode": False,
+    "external": False,
```
[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/acq/fastem.py:172]: No, overview and live stream are both using scan mode, so external = False

[the remark this answers is not in the document]
Comment (Sabrina85) [reply src/odemis/acq/fastem.py:172]: `# Full frame` is correct - sorry.

**Comment by philipwinklerdelmic** [comment]
Ok, it's getting a bit complicated
@tepals Can you have a look at the xt commit? I think there might be a bug with the autoblanking functionality, see commit message. I propose to completely change it and use an additional VA. But maybe I'm wrong.

Other than that, I moved the configure_scanner function to a different location to avoid ciruclar dependencies in the live stream. The livestream needs this function in .prepare to set up the correct values.

If everyone agrees with the changes, I need to clean up a bit.

@Sabrina85 why don't we use autoblanking on the megafield acquisition?

**Comment by tepals** [comment]
@philipwinklerdelmic I think using an autoblank VA is fine. I based the current implementation on the Phenom driver, but that does not have an update settings function, so the blanker VA is not overwritten.

**Review by pieleric — CHANGES_REQUESTED** [review]
Just some minor comments about the test case.

Once you've fixed them, it'd be nice if you could rewrite the history to have (at least) two commits: one for the xt_client changes, and one for the actual fastem_conf.

**Comment by Sabrina85** [comment]
@pieleric Can you have a last look on the last commit please? Afterwards, I will squash everything.

**Review by pieleric — APPROVED** [review]
Looks fine to me. Could you squash into 2 commits?

**Comment by Sabrina85** [comment]
> Could you squash into 2 commits?

Done

**Pull request merged**